#summary Prediction of transcription factor binding sites using HMMs on DNase-seq and histone modification ChIP-seq data.

= Footprints =

Command: *rgt-footprint*

== Method ==

We propose an HMM-based approach to integrate both DNase I hypersensitivity and histone modifications for the detection of open chromatin regions and active binding sites.

Within transcription factor binding sites, there is a specific grammar of DNase I digestion and histone marks. We have therefore devised a multivariate HMM to model this regulatory grammar by simultaneous analysis of DNase-seq and the ChIP-seq profiles of H3K4me3 (indicative of promoters) or H3K4me1 (indicative of enhancers) on a genome-wide level.

The HMM has as input a normalized and a slope signal of DNase-seq and one of the histone marks. It can therefore detect the increase, top and decrease regions of either histone modification and DNase signals. The genomic regions annotated with the 'footprint' HMM state are considered our predictions and represent likely binding sites within that cell's context.

== Requirements ==

This tool requires a few extra Python packages:
  * [http://www.numpy.org/ numpy]
  * [http://www.scipy.org/ scipy] 
  * [http://scikit-learn.org/stable/ scikit]
  * [http://code.google.com/p/pysam/ pysam]

If you have sudo privileges, the first three packages from the list above can be installed with the following commands:
{{{
sudo apt-get install build-essential python-dev python-numpy python-setuptools python-scipy libatlas-dev libatlas3-base
pip install -U scikit-learn
}}}

However, if you are not the administrator. Please visit the project's websites by clicking in the links above and search for the python installation with the "prefix" scheme.

== Installation ==

Given that you have all the required packages. The tool can be easily installed locally by executing the following commands:
{{{
# Installing subversion (SVN)
apt-get install subversion

# Downloading code and change to the code's directory.
svn checkout http://reg-gen.googlecode.com/svn/trunk/ ~/reg-gen/
cd ~/reg-gen/

# Setting up environmental variables for local installation
# Please also add the following two lines to your ~/.bashrc file.
export PATH=$PATH:~/app/bin/
export PYTHONPATH=$PYTHONPATH:~/app/lib/pythonX.X/:~/app/lib64/pythonX.X/

# Installation command
python setup.py install --prefix=~/app --rgt-tool=footprint
}}}

For more details on the installation please go to the [https://code.google.com/p/reg-gen/wiki/Installation installation] section.

== Input ==

==== Required Input: ====
|| `-dnase_file=<FILE>` || File containing the DNase-seq signal. The file consists of the genomic coverage (read overlap) of the aligned reads. A high-resolution signal should be given correspoding to the overlap of all the 5' read base pairs.<br>Format: bigwig.<br> Default: None.||
|| `-histone_file=<FILE>` || File containing the ChIP-seq signal from histone modifications. The file consists of the genomic coverage of the aligned reads. Each read should be extended to the average length of immunoprecipitated fragments.<br>Format: bigwig.<br>Default: None.||
|| `-hmm_file=<FILE>` || File containing the HMM parameters. Check the manual for more information on the format and how to obtain these models.<br>Format: hmm.<br>Default: None.||

==== Optional Input: ====
|| `-coord_file=<FILE>` || Coordinate file containing the regions of the genome where the footprinting will be performed. If None, the footprinting is applied to the complete genome.<br>Format: bed.<br>Default: Complete genome tiled in 10000bp regions.||


==== Input Parameters: ====
|| `-organism=<STRING>` || Organism considered on the analysis. Can be 'hg19'  or 'mm9'. All the default files are going to be based on the chosen organism.<br>Default: hg19||


==== Output Options: ====
|| `-output_location=<PATH>` || Path where the output files will be written.<br>Default: current directory.||

== Input Format Definition ==

=== Experimental Matrix ===

Under construction.

=== HMM File ===

Under construction.

== Output ==

Under construction.

== Example ==

Under construction.