#summary Future functionalities and bug correction.

= TO-DO =

== GenomicRegion ==

|| *Developer* || *Function* || *Input* || *Output* || *Description* ||
|| || overlap() || || || It doesn't work for the zero-length region. For example, (initial=20, finial=20) with the same region ||

== GenomicRegionSet ==

|| *Developer* || *Function* || *Input* || *Output* || *Description* ||
|| || read_bigbed || - || - || Create self based on bigbed (.bb) format. ||
|| || write_bigbed || - || - || Writes self in bigbed (.bb) format. ||
|| || write_fasta || - || - || Writes self in fasta (.fa) format. ||
|| JK || intersect() || || || Correct bug (talk to EG). ||
|| JK || merge() || || || Correct bug (talk to EG). ||
|| JK || random_regions() || || || Put flag for chrY ||
|| EG || gene_association() || || || 1. Implement multiple gene associations (more than 2) <br> 2. Treat the gene regions file in order to merge all entries with same gene name that overlap. ||
|| || || || || ||

== CoverageSet ==

|| *Developer* || *Function* || *Input* || *Output* || *Description* ||
|| || coverage_from_bed || - || - || Create self based on bed format. ||
|| || coverage_from_bigbed || - || - || Create self based on bigbed (.bb) format. ||
|| || coverage_from_wig || - || - || Create self based on wig format. ||
|| || coverage_from_bigwig || - || - || Create self based on bigwig (.bw) format. ||
|| || write_bigbed || - || - || Writes self in bigbed (.bb) format. ||
|| || || || || ||

== ExperimentalMatrix ==

|| *Developer* || *Function* || *Input* || *Output* || *Description* ||
|| JK ||  || || || Add verbose flag to prevent writing status in sys.err. || || || || || ||

== GeneSet ==

|| *Developer* || *Function* || *Input* || *Output* || *Description* ||
|| || || || || ||

== GeneMapping ==

|| *Developer* || *Function* || *Input* || *Output* || *Description* ||
|| || || || || ||

== Util ==

|| *Developer* || *Function* || *Input* || *Output* || *Description* ||
|| || || || || ||

== Tools ==

=== associationAnalysis ===
  * include the coverageFromBed function and box plot function
  * include the geneCogerageFromBed tool and box plot function

=== footprint ===

=== motifanalysis ===
  * EG - Make heatmap
  * EG - Make network
  * EG - Full profiling (time and memory)
  * EG - Motif clustering (see HOCOMOCO and consider tomtom)
  * EG - Keep original gene names for GO search. [low]

== Data ==

  * EG - Include UCSC binaries.